Sage200tests

tyler-cloud-elements opened the pull request:
## Highlights
* Sage200 finance element (develop-CE-3504)
* Churros integration tests (churros/sage200tests) for banks, customers, invoices, payments, products and transactions.

## Notes
* Configuration values for callback_url, client_id, client_secret and scope are all the public development values for Sage200. I’m still waiting on hearing back from Sage about my application to get our own ClientId/Secret/Scope and I will modify these accordingly. 
* POSTs to /customers has a very small margin for failure, due to the sheer amount of customers I've had to create for various payments (can't delete a customer with a payment/invoice/etc.). So I used a random number generator, which _occasionally_ will repeat a used customer's reference number. Rerunning the test should move it from fail -> pass.

jjwyse commented on src/test/elements/sage200/assets/transformations.json:
```diff
@@ -0,0 +1,95 @@
+{
+  "customers": {
+    "vendorName": "https://api.columbus.sage.com/uk/sage200/accounts/v1/customers",
```
should this really be the `vendorName` ? :confused:

jjwyse commented on src/test/elements/sage200/banks.js:
```diff
@@ -0,0 +1,12 @@
+const cloud = require('core/cloud');
+
+suite.forElement('finance', 'banks', null, (test) => {
+  it('Should support GET', () => {
```
i think you can just do `test.should.return200OnGet()` here instead of needing the `it` block with the `cloud` call

jjwyse commented on src/test/elements/sage200/customers.js:
```diff
@@ -0,0 +1,23 @@
+
+suite.forElement('finance', 'customers', payload, (test) => {
+  let customerId;
+  it('Should support CRUDS with custom PATCH', () => {
```
@bscharm - with your new changes to support custom update JSON, can we replace that with that functionality?

bscharm replied:
@jjwyse @tyler-cloud-elements yes should be able to use that in this case.
```
const options = {
  churros: {
    updatePayload: customerUpdate()
  }
};

suite.forElement('finance', 'customers', payload, (test) => {
  test.withOptions(options).should.supportCruds();
});
```

jjwyse commented on src/test/elements/sage200/invoices.js:
```diff
@@ -0,0 +1,35 @@
+      .then(r => cloud.get(test.api),{qs: {where: 'urn=\''+urnId +'\''}})
+      .then(r => invoiceId = r.body[0].id)
+      .then(r => cloud.get(test.api + '/' + invoiceId));
+      //can't delete a customer with transactions ...
```
can we remove commented out code?

jjwyse commented:
@tyler-cloud-elements - one other thing - i noticed you mentioned that these will randomly fail due to the random generator however our entire suite of tests now gets run on every merge into `develop` and, in the near future, the entire `churros` suite will get run on every single pull request, in which case a pull request will only be permitted to be merged *if* all tests pass.  it is very important we do our best not to write fragile tests so is there any way you can make the above tests more reliable so we're not gettin false negatives?

jjwyse commented:
@tyler-cloud-elements - hey tyler, going to be out today and tomorrow with no access to my computer, so if you get the above changes made before then, can you assign this to @bscharm to look over?  thanks!

tyler-cloud-elements commented:
* Deleted a duplicate sharepoint case in Oauth.js
* Addressed code review concerns
* As for the random test fail issue, it shouldn't be a problem. When Atul gets back in town, we can clean up all the excess customers through their windows UI application.

jjwyse commented:
@tyler-cloud-elements - sorry, one last thing before we merge this in.  can you make the `it(...)` test names (i.e. `Should ...`) lowercased to be consistent with all of our other tests?  I know, I'm OCD and super picky :/ haha sorry.

tyler-cloud-elements commented:
Addressed code review concerns

jjwyse commented:
great, thanks @tyler-cloud-elements, nice work!  will merge this in once element PR is merged into platform.
